brave/src/main/java/brave/internal/propagation/CorrelationFieldScopeDecorator.java

```diff
@@ -13,85 +13,184 @@
+import java.util.LinkedHashSet;
+import java.util.Locale;
+import java.util.Set;
 
 /**
- * Adds correlation properties "traceId", "parentId", "spanId" and "sampled" when a {@link
- * brave.Tracer#currentSpan() span is current}. "traceId" and "spanId" are used in log correlation.
+ * Adds correlation properties "traceId", "parentId", "spanId" and "sampled" when a {@linkplain
+ * Tracer#currentSpan() span is current}. "traceId" and "spanId" are used in log correlation.
  * "parentId" is used for scenarios such as log parsing that reconstructs the trace tree. "sampled"
  * is used as a hint that a span found in logs might be in Zipkin.
  */
 public abstract class CorrelationFieldScopeDecorator implements ScopeDecorator {
+  final Updater[] updaters;
+
+  protected static abstract class Builder<B extends Builder<B>> {
+    final Set<String> extraFields = new LinkedHashSet<>(); // insertion order;
+
+    /**
+     * Adds a field from {@link ExtraFieldPropagation} into the correlation context.
+     *
+     * @since 5.11
+     */
+    public B addExtraField(String name) {
+      if (name == null) throw new NullPointerException("name == null");
+      String lowercase = name.toLowerCase(Locale.ROOT); // contract of extra fields internally
+      extraFields.add(lowercase);
+      return (B) this;
+    }
+
+    public abstract CurrentTraceContext.ScopeDecorator build();
+  }
+
+  protected CorrelationFieldScopeDecorator(Builder<?> builder) {
+    String[] extraFields = builder.extraFields.toArray(new String[0]);
+    updaters = new Updater[4 + extraFields.length];
+    updaters[0] = new TraceIdUpdater(this);
+    updaters[1] = new ParentSpanIdUpdater(this);
+    updaters[2] = new SpanIdUpdater(this);
+    updaters[3] = new SampledUpdater(this);
+    for (int i = 0; i < extraFields.length; i++) {
+      updaters[4 + i] = new ExtraFieldUpdater(this, extraFields[i]);
+    }
+  }
 
   /**
    * When the input is not null "traceId", "parentId", "spanId" and "sampled" correlation properties
    * are saved off and replaced with those of the current span. When the input is null, these
    * properties are removed. Either way, "traceId", "parentId", "spanId" and "sampled" properties
    * are restored on {@linkplain Scope#close()}.
    */
-  @Override public Scope decorateScope(@Nullable TraceContext currentSpan, Scope scope) {
-    String previousTraceId = get("traceId");
-    String previousSpanId = get("spanId");
-    String previousParentId = get("parentId");
-    String previousSampled = get("sampled");
-
-    if (currentSpan != null) {
-      maybeReplaceTraceContext(
-        currentSpan, previousTraceId, previousParentId, previousSpanId, previousSampled);
-    } else {
-      remove("traceId");
-      remove("parentId");
-      remove("spanId");
-      remove("sampled");
+  @Override public Scope decorateScope(@Nullable TraceContext context, Scope scope) {
+    String[] previousValues = new String[updaters.length];
```

Comment:
using a mutable array here is intentional. It allows a later change to implement immediate flushing https://github.com/openzipkin/brave/issues/682

Reply:
Not completely sure but I thought the immediate flushing would mean there are extra fields added, so doesn't it need to be expandable like a `List`?

Reply:
actually the late thing is change in a value not change in the whitelist. so should be fine. this will be separate change anyway

Reply:
yeah this is just late updates keys, set in the builder and immutable after. so the value holder needn't be expandable. If we change later to make this expandable, we would need to update this code, as would other code which assumes the list is fixed.